BJS2-16335 Projects

fokinms opened the pull request:
Projects

fokinms committed:
Create ProjectRepository with custom queries
Create ProjectController & ProjectService with methods:
-create
-update
-findProjectsWithFilter
-findAllProjects
-findById
Create ProjectMapper
Create ProjectDto & ProjectFilterDto
Create ProjectServiceValidator
Create ProjectFilter with filter classes
Create ValidationException

fraifree commented on src/main/java/faang/school/projectservice/controller/project/ProjectController.java:
```diff
@@ -0,0 +1,55 @@
+
+    private final ProjectService projectService;
+
+    @PostMapping("{userId}/projects/create")
```
Нужно подумать над url-ками для методов, у тебя и сам контроллер projects и в урлах тоже слово висит, если ты к примеру создаешь проект, то можно написать так: "/create", еще нюанс, ты уже в дто передаешь ownerId, зачем тебе передавать эту информацию 2 раза, если можно оттуда достать?

fokinms replied:
Done

fraifree commented on src/main/java/faang/school/projectservice/controller/project/ProjectController.java:
```diff
@@ -0,0 +1,55 @@
+        return projectService.create(userId, projectDto);
+    }
+
+    @PutMapping("/projects")
```
Мне кажется стоит переименовать

fokinms replied:
Done

fraifree commented on src/main/java/faang/school/projectservice/controller/project/ProjectController.java:
```diff
@@ -0,0 +1,55 @@
+        return projectService.update(projectDto);
+    }
+
+    @PostMapping("/projects/{userId}/filtered")
```
Мне кажется стоит переименовать

fokinms replied:
Done

fraifree commented on src/main/java/faang/school/projectservice/controller/project/ProjectController.java:
```diff
@@ -0,0 +1,55 @@
+        return projectService.findProjectsWithFilter(userId, projectFilterDto);
+    }
+
+    @GetMapping("/projects/{userId}/projects")
```
Мне кажется стоит переименовать, тут у тебя в пути слово project 3 раза будет

fokinms replied:
Done

fraifree commented on src/main/java/faang/school/projectservice/controller/project/ProjectController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @GetMapping("/projects/{userId}/projects")
+    public List<ProjectDto> findAllProjects(@Valid @PathVariable Long userId) {
```
В задании у тебя получить все проекты, на счет пользователя там ничего не указано, мне кажется поле userId тут лишнее

fokinms replied:
Оставил как есть на ревью. В задании есть дополнительная логика по ограничению доступа к приватным проектам, для ее реализации нужен userId

fraifree commented on src/main/java/faang/school/projectservice/controller/project/ProjectController.java:
```diff
@@ -0,0 +1,55 @@
+        return projectService.findAllProjects(userId);
+    }
+
+    @PostMapping("/projects/{userId}/project")
```
Думаю стоит переделать url

fraifree commented on src/main/java/faang/school/projectservice/controller/project/ProjectController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @PostMapping("/projects/{userId}/project")
+    public  ProjectDto findById(@Valid @PathVariable Long userId,
```
Этот метод ишет у тебя проект по Id, вопрос о целесообразности передачи сюда id usera

fokinms replied:
видимость проекта для делающего запрос в зависимости от visibility и участия запрашивающего в проекте

fraifree commented on src/main/java/faang/school/projectservice/filter/project/ProjectIdFilter.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.stream.Stream;
+
+@Component
+public class ProjectIdFilter implements ProjectFilter {
```
Не совсем понимаю, зачем создавать фильтр по Id, это уникальный идентификатор, у тебя в задании есть метод для его поиска, по id ты в любом случае получишь только 1 результат

fokinms replied:
Оставил как есть на ревью. В задании есть дополнительная логика по ограничению доступа к приватным проектам, для ее реализации нужен этот фильтр

InnMish replied:
Посмотрю логику, но пока не поняла все еще...))

fraifree commented on src/main/java/faang/school/projectservice/dto/project/ProjectFilterDto.java:
```diff
@@ -0,0 +1,21 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class ProjectFilterDto {
+    private Long idPattern;
```
Есть ли смысл в такой фильтрации?

fokinms replied:
На мой взгляд по заданию есть. комментарий чуть выше

InnMish replied:
А зачем? У проекта есть только 1 айди 
как тут фильтровать? Это по сути фильтр get-запрос по айди)

fokinms replied:
У нас есть один эндпоинт, который ищет проект и по юзер айди, и по айди проекта. у проектов есть статус visibility, и его надо учитывать при поиске по заданию. айди юзера из строки запроса беру, а айди проекта из ДТО фильтра, поэтому он тут есть)

fraifree commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,98 @@
+        return projectMapper.toDto(project);
+    }
+
+    @Transactional
```
Думаю стоит добавить readOnly = true

fokinms replied:
Done

fraifree commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,98 @@
+        return projectMapper.toDtos(filteredProjects);
+    }
+
+    @Transactional
```
Думаю стоит добавить readOnly = true

fokinms replied:
Done

fraifree commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,98 @@
+                .toList());
+    }
+
+    @Transactional
```
Думаю стоит добавить readOnly = true

fokinms replied:
Done

fokinms committed:
fix cross-review comments and some validation logic

InnMish commented on src/main/java/faang/school/projectservice/controller/project/ProjectController.java:
```diff
@@ -0,0 +1,55 @@
+
+    private final ProjectService projectService;
+
+    @PostMapping("{userId}/create")
```
https://discord.com/channels/1102662981743030334/1102662982447669361/1265711557434675231

fokinms replied:
Ок

InnMish commented on src/main/java/faang/school/projectservice/controller/project/ProjectController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @PostMapping("/{userId}/accessible/filtered")
+    public List<ProjectDto> findProjectsWithFilter(@Valid @PathVariable Long userId,
```
а что делает Valid на Long поле?)

InnMish commented on src/main/java/faang/school/projectservice/filter/project/ProjectDescriptionFilter.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Stream<Project> apply(Stream<Project> projects, ProjectFilterDto filters) {
+        return projects.filter(project -> project.getDescription().equals(filters.getDescriptionPattern()));
```
я бы тут по contains смотрела 
есть же вероятность, что описание не один в один совпадет 
знаешь, как на маркетплейсе ищешь по тегам в описании) Можно и тут так сделать

fokinms replied:
Done

InnMish commented on src/main/java/faang/school/projectservice/dto/project/ProjectDto.java:
```diff
@@ -0,0 +1,30 @@
+    private String description;
+    private Long ownerId;
+    private List<Long> childrenId;
+    private List<Long> tasksId;
```
taskIds только

InnMish commented on src/main/java/faang/school/projectservice/dto/project/ProjectDto.java:
```diff
@@ -0,0 +1,30 @@
+    private Long ownerId;
+    private List<Long> childrenId;
+    private List<Long> tasksId;
+    private LocalDateTime created_at;
```
в джаве нейминг полей без нижнего подчеркивания 
createdAt

InnMish commented on src/main/java/faang/school/projectservice/service/project/validation/ProjectServiceValidator.java:
```diff
@@ -0,0 +1,38 @@
+package faang.school.projectservice.service.project.validation;
```
пакет validator вынеси на один уровень с сервисом 
и класс назови просто ProjectValidator  

1. пакет - унифицированное название, не процесс, а ответственный компонент - сервис, контроллер, валидатор 
2. Не разносим валидации котроллеров и сервисов на отдельные класс, валидируем именно сущность

fokinms replied:
Класс переименовал и перенес, уже замердженная папка называется validation. Оставим так или переименуем?

InnMish commented on src/main/java/faang/school/projectservice/service/project/validation/ProjectServiceValidator.java:
```diff
@@ -0,0 +1,38 @@
+
+    public void validateCreation(Long userId, ProjectDto projectDto) {
+        if (projectRepository.findAll().stream()
+                .filter(project -> project.getOwnerId().equals(userId))
```
Я бы вот тут не проверяла совпадение, а userId сеттила в ownerId. Тут видишь, userId - это как авторизационный идентификатор пользователя, который мы будем получать из мета-данных x-user-id (ты кстати уже это можешь сделать, на платформе вся инва есть). И таким образом мы знаем, кто создает проект. Поэтому можем не заставлять пользователя в дто себя класть, а сами это сделать на бэке

fokinms replied:
здесь userId из параметра запроса, а имя проекта беру дальше из ДТО. Валидирую по заданию, что у пользователя нет уже существующих проектов с таким именем. Т.е. фильтр в валидаторе точно по заданию нужен. насчет x-user-id - начну использовать, он во всех проектах есть?. В целом, этот метод можно так и оставить, параметры эти все равно нужны, передаются из сервиса

InnMish replied:
Да, есть во всех основных сервисах, чтобы feign использовать) 
Ты можешь с ним разобраться для дальнейших задач

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private void entityCreatePreparation(Long userId, Project project) {
+        project.setOwnerId(userId);
```
так мы же в валидаторе проверяли.. что owner есть

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,99 @@
+    public ProjectDto create(Long userId, ProjectDto projectDto) {
+        projectServiceValidator.validateCreation(userId, projectDto);
+        Project project = projectMapper.toEntity(projectDto);
+        entityCreatePreparation(userId, project);
```
этот метод у тебя один раз только используется вот в этом методе и все, а ты в нем по сути просто сеттишь поля. я бы тут прям это сделала.. там буквально две строчки будет.. но это мелочь, смотри сам

fokinms replied:
Согласен. Переделал

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Transactional
+    public ProjectDto update(ProjectDto projectDto) {
+        projectServiceValidator.validateUpdating(projectDto);
```
я вот знаешь, что думаю 
со стороны бизнес-логики 
если у нас проект такой раньше был, то он как бы в базе и ок, мы его обновим методом save. 
Ты пользователю нигде не даешь возможность именно удалить проект из бд, соответственно, его можно только закрыть, но в бд он останется. Поэтому такая валидация нам не очень много смысла дает как будто. Я бы лучше проверила, что проект не закрыт)

fokinms replied:
Переделал)

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,99 @@
+
+    private void entityUpdatePreparation(Project project) {
+        if (checkUpdateReason(project)) {
+            project.setUpdatedAt(LocalDateTime.now());
```
охохо эхехе 
смотри 
в сущность Project над этим полем висит аннотация UpdatingTimestamp 
Благодаря ней у нас дата в бд атоматически обновляется и тут ее не нужно руками сеттить

fokinms replied:
Да. На момент написания еще не знал) Текст задания немного спутывает. Говорит что надо обратить внимание на установку времени. Возможно это и хорошо, сделал ошибку - лучше запомнится)

fokinms committed:
Merge branch 'refs/heads/griffon-master-bc5' into fms-BJS2-16335

fokinms committed:
fix after review

fokinms committed:
Fix comments
Fix ProjectService and add method comment
Fix ProjectValidator and add new method
Create ProjectValidatorTest
Create ProjectServiceTest
Create Filter tests

fokinms committed:
Merge branch 'refs/heads/griffon-master-bc5' into fms-BJS2-16335

# Conflicts:
#	src/main/java/faang/school/projectservice/repository/ProjectRepository.java
#	src/main/java/faang/school/projectservice/validator/ProjectValidator.java
#	src/test/java/faang/school/projectservice/validator/ProjectValidatorTest.java

fokinms committed:
Fix comments

Commit:
deleted duplicated validator

Commit:
springboottest deleted
